python/gstaichi/lang/_fast_caching/src_hasher.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, Iterable, Sequence
+
+from pydantic import BaseModel
+
+from . import args_hasher, config_hasher, function_hasher
+from .fast_caching_types import FunctionSourceInfo, HashedFunctionSourceInfo
+from .hash_utils import hash_string
+from .pyside_cache import PysideCache
+
+
+def create_cache_key(kernel_source_info: FunctionSourceInfo, args: Sequence[Any]) -> str | None:
+    """
+    cache key takes into account:
+    - arg types
+    - cache value arg values
+    - kernel function (but not sub functions)
+    - compilation config (which includes arch, and debug)
+    """
+    args_hash = args_hasher.hash_args(args)
+    if args_hash is None:
+        return None
+    kernel_hash = function_hasher.hash_kernel(kernel_source_info)
+    config_hash = config_hasher.hash_compile_config()
+    cache_key = hash_string(f"{kernel_hash}_{args_hash}_{config_hash}")
+    return cache_key
+
+
+class CacheValue(BaseModel):
+    hashed_function_source_infos: list[HashedFunctionSourceInfo]
+
+
+def store(cache_key: str, function_source_infos: Iterable[FunctionSourceInfo]) -> None:
+    """
+    Note that unlike other caches, this cache is not going to store the actual value we want. This cache is only used for verification
+    that our cache key is valid. Big picture:
+    - we have a cache key, based on args and top level kernel function
+    - we want to use this to look up LLVM IR, in C++ side cache
+    - however, before doing that, we first want to validate that the source code didn't change
+        - i.e. is our cache key still valid?
+    - the python side cache contains information we will use to verify that our cache key is valid
+        - ie the list of function source infos
+    """
+    if not cache_key:
+        return
+    cache = PysideCache()
+    hashed_function_source_infos = function_hasher.hash_functions(function_source_infos)
+    cache_value_obj = CacheValue(hashed_function_source_infos=list(hashed_function_source_infos))
+    cache_value_json = cache_value_obj.json()
```

Comment:
I would avoid this temporary as it is already perfectly explicit, but as you wish.